src/secureApi.js

```diff
@@ -0,0 +1,462 @@
+// Copyright 2015-2017 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+import uniq from 'lodash/uniq';
+import store from 'store';
+
+import Api from '@parity/api';
+import { LOG_KEYS, getLogger } from '@parity/shared/lib/config';
+
+const log = getLogger(LOG_KEYS.Signer);
+
+const RPC_URL = '127.0.0.1:8545';
+const WS_URL = '127.0.0.1:8546';
+const UI_URL = '127.0.0.1:8180';
+
+export default class SecureApi extends Api {
+  _isConnecting = false;
+  _needsToken = false;
+  _tokens = [];
+  _uiApi = null;
+
+  _dappsUrl = null;
+  _wsUrl = null;
+  _url = null;
+
+  static getHttpProvider(url, protocol) {
+    return new Api.Provider.Http(`${protocol}//${url || RPC_URL}/rpc`, 0);
+  }
+
+  static getWsProvider(url, protocol, sysuiToken) {
+    const transportUrl = SecureApi.transportWsUrl(url || WS_URL, protocol);
+
+    return new Api.Provider.Ws(transportUrl, sysuiToken, false);
+  }
+
+  static transportWsUrl(url, protocol) {
+    const proto = protocol() === 'https:' ? 'wss:' : 'ws:';
+
+    return `${proto}//${url}`;
+  }
+
+  // Returns a protocol with `:` at the end.
+  static protocol() {
+    return window.location.protocol === 'file:'
+      ? 'http:'
+      : window.location.protocol;
+  }
+
+  constructor(
+    uiUrl,
+    nextToken,
+    getProvider = SecureApi.getWsProvider,
+    protocol = SecureApi.protocol
+  ) {
+    const sysuiToken = store.get('sysuiToken');
+
+    // TODO
+    // Connecting to WS_URL
+    // Should connect to uiUrl to get an elevated API
+    // -Amaury 18.01.2018
+    const wsProvider = getProvider(null, protocol, sysuiToken);
+    // const wsProvider = getProvider(uiUrl, protocol, sysuiToken);
+
+    super(wsProvider);
+
+    this.protocol = protocol;
+    this._url = uiUrl || UI_URL;
+
+    // TODO
+    // Connecting to RPC_URL
+    // Should connect to uiUrl to get an elevated API
+    // -Amaury 18.01.2018
+    const httpProvider = SecureApi.getHttpProvider(null, this.protocol());
+    // const httpProvider = SecureApi.getHttpProvider(this._url, this.protocol());
```

Comment:
Idem here, connecting to RPC_URL

Reply:
This will be an issue when people override the default settings - the way around it here will most probably be to give command-line args to the shell. (Since it cannot figure things out by itself)